README.md

```diff
@@ -1,229 +1,894 @@
+you'll also see two more pieces of information: function latency (e.g., 73
+microseconds for `__x64_sys_bpf`) and it's return result, which could be one of
+`-Exxx` error codes, `NULL`, `true`/`false`, some valid address like
+`0xffffc90000243000`, some other integer value, or even `void`; `retsnoop` will
+do its best to represent captured return value as meaningfully as possible,
+based on captured return value and function's type signature, if available. In
+our case, we see that innermost call to `bpf_map_alloc_percpu` returned `NULL`,
+which the got translated to `-ENOMEM` error and bubbled up all the way through
+`__x64_sys_bpf` and eventually got returned to user-space as a result of
+`bpf()` syscall.
+
+There could be few extra markers like '!', '*', etc, at the very left side of a
+line, but we won't go into detail on what they mean. Usually it's pretty
+harmless to ignore them.
+
+So, in general, what did we achieve and what can we understand from
+`retsnoop`'s output? With a single invocation of `retsnoop` with few arguments
+we managed to trace `bpf()` syscall all the way down to `bpf_map_alloc_percpu`
+function returning NULL. If we dig deeper into source code, we'll see that it
+further calls `__alloc_percpu_gfp` (which we didn't instruct `retsnoop` to
+trace, as it wasn't in either `-e` nor `-a` arguments, so we can't see that
+from the above stack trace) which got to be the one further returning `NULL`,
+indicating allocation failure. We can keep digging deeper to try to understand
+exactly why, by reading code and adjusting `retsnoop`'s set of traced function
+, but that's beyond the scope of this example.
+
+The main point is that we can see the sequence of function calls
+leading all the way to an attempt to allocate memory for per-CPU BPF array map,
+and that it is exactly `bpf_map_alloc_percpu` that's failing. We also know that
+`NULL` got translated into `-ENOMEM` before bubbling up to user-space as an
+error returned from `bpf()` syscall.  Without `retsnoop` it would be a
+considerably harder thing to narrow it down so precisely without the need to
+modify, recompile, and install new kernel version.
+
+Note also that we can see how did we get to our entry function
+(`__x64_sys_bpf`) through a chain of `entry_SYSCALL_64_after_hwframe` ->
+`do_syscall_64` -> `__x64_sys_bpf` calls, even though we didn't instruct
+`retsnoop` to trace first two functions. This ability to see how we got to our
+entry function without explicitly specifying possible functions with `-e` or
+`-a` is extremely useful, especially when dealing with unfamiliar parts of the
+kernel.
+
+## Tracing BPF verification flow
+
+`retsnoop` was initially designed to debug cases when kernel returns some
+generic error (like `-EINVAL`) from somewhere deep inside complex syscall (like
+`bpf()` syscall) and it's hard to guess that that error is. So by default
+`retsnoop` will only catch call stack that end up returning error (`-Exxx` or
```

Comment:
s/catch call stack that end up returning error/catch the call stack that ends up returning an error